### DISCUSSION TITLE
New rule: needless-foreach, or ban-foreach

### DISCUSSION THREAD

**Discussion by jpike88** [rome/tools#4331] [Suggestions]
Foreach is an anti pattern as it creates more difficult to read code. There's no good reason to use it. It should be banned, and part of the Rome recommended ruleset.

**Comment by Conaclos** [comment]
Any pointer to justify it?

**Reply by jpike88** [reply]
It creates harder to read code, with zero advantage. It’s semantically more consistent to use a proper for of statement for arrays because we already do it for other iterative objects.

**Reply by GabenGar** [reply]
There is no reason to use `for of` where `Array.forEach()` would suffice. The latter declares the intent - an iteration with side effects over entirety of array, while the former can mean anything and requires squinting at the body to infer intent.
If anything there should be an opposite rule - ban `for of` where `forEach()` is enough.

**Reply by jpike88** [reply]
There are a few good reasons:

- Code readability and semantics can be debated endlessly because it has a degree of subjectivity, so not going to press my opinion there, but I much prefer the more language agnostic use of code blocks versus using a function with a callback in it, purely just to iterate an array. In our (very big) codebase, all we ever use `for of` on are arrays. The principle I like to follow is the less callbacks, the less cognitive overhead required to parse the code. The fact that a simple loop can look structurally similar to a map, reduce, filter function just doesn't sit will with me.
- `forEach` is over 2 times slower than `for of`
- Finally, it creates inconsistent coding styles as some developers will just prefer using `forEach` all the time versus others, and makes the code more inconsistent.

![forEach](https://user-images.githubusercontent.com/9585787/230767874-6b04c78a-d772-47bf-9a52-57db80ebbd45.svg)
taken from https://leanylabs.com/blog/js-forEach-map-reduce-vs-for-for_of/

**Reply by GabenGar** [reply]
> purely just to iterate an array

Yeah, and there are array methods exactly for that.

>In our (very big) codebase, all we ever use for of on are arrays.

The logic behind the callback doesn't suddenly vaporize in `for of` constructs, it just moves them into the iterator body, where it exists in the same scope the iteration happens, as opposed to a separate scope of the callback. 

> The principle I like to follow is the less callbacks, the less cognitive overhead required to parse the code.

It's no "less overhead", it's just spilling the logic of the iteration into the parent scope and polluting its namespace with extra symbols.

> The fact that a simple loop can look structurally similar to a map, reduce, filter function just doesn't sit will with me.

But all `for of` loops looking similar despite covering an even wider range of iteration logic (down to having a custom iterator) is completely okay?

> forEach is over 2 times slower than for of

And you tested this in your (very big) codebase, not just confirming your beliefs with pictures from a blog, right?

> Finally, it creates inconsistent coding styles as some developers will just prefer using forEach all the time versus others, and makes the code more inconsistent.

`for of` loops can't be consistent at all by their nature.

You are being disingenuous with that graph, as this is a perf grap for a side-effect-free `for of` iteration (which happens never in real code) and therefore trivially optimizable by interpreter. Notice how it drops sharply when it does an equivalent of `map()`, aka introducing a single side effect. Since they didn't test the performance with ever increasing amount of side effects, I'll just assume `for of` loops become progressively worse with them and the only optimizations possible to do are by magical compiler, hence why any sufficiently complex `for of` loop always becomes an unmanageable mess with no way for decomposition.

**Reply by jpike88** [reply]
Let’s agree to disagree. I consider this rule to be highly valuable and I’m sure there are others who feel that way as well. You simply can choose not to enable it.

**Reply by ematipico** [reply]
If a volunteer wants to implement this rule, I don't have anything against it, but I wouldn't consider it a rule to recommend.

I would consider this rule a low priority and I would advise that there are more important rules to implement other than this one.

**Reply by jpike88** [reply]
forgot to post relevant eslint equivalent https://github.com/github/eslint-plugin-github/blob/main/docs/rules/array-foreach.md

**Reply by ematipico** [reply]
> forgot to post relevant eslint equivalent https://github.com/github/eslint-plugin-github/blob/main/docs/rules/array-foreach.md

That's a bit misleading, though. That's not an eslint rule, but a rule recommended by a plugin.

**Reply by jpike88** [reply]
My bad, I meant "eslint 'rule' equivalent", as in a rule developed to be used in eslint via a plugin.

**Reply by jpike88** [reply]
closing as it has been implemented. thanks for your hard work!
